unified_planning/io/pddl_writer.py

```diff
@@ -145,72 +182,101 @@ class PDDLWriter:
-            if problem_kind.has_existential_conditions():
+            if self.problem_kind.has_existential_conditions():
                 out.write(' :existential-preconditions')
-            if problem_kind.has_universal_conditions():
+            if self.problem_kind.has_universal_conditions():
                 out.write(' :universal-preconditions')
-            if (problem_kind.has_continuous_time() or
-                problem_kind.has_discrete_time()):
+            if (self.problem_kind.has_continuous_time() or
+                self.problem_kind.has_discrete_time()):
                 out.write(' :durative-actions')
-            if problem_kind.has_duration_inequalities():
+            if self.problem_kind.has_duration_inequalities():
                 out.write(' :duration-inequalities')
-            if (self.problem.kind.has_actions_cost() or
-                self.problem.kind.has_plan_length()):
+            if (self.problem_kind.has_actions_cost() or
+                self.problem_kind.has_plan_length()):
                 out.write(' :action-costs')
             out.write(')\n')
 
-
-        if problem_kind.has_hierarchical_typing():
-            while self.problem.has_type(self.object_freshname):
-                self.object_freshname = self.object_freshname + '_'
+        if self.problem_kind.has_hierarchical_typing():
             user_types_hierarchy = self.problem.user_types_hierarchy
             out.write(f' (:types\n')
             stack: List['unified_planning.model.Type'] = user_types_hierarchy[None] if None in user_types_hierarchy else []
-            out.write(f'    {" ".join(self._type_name_or_object_freshname(t) for t in stack)} - object\n')
+            out.write(f'    {" ".join(self._get_mangled_name(t) for t in stack)} - object\n')
             while stack:
                 current_type = stack.pop()
                 direct_sons: List['unified_planning.model.Type'] = user_types_hierarchy[current_type]
                 if direct_sons:
                     stack.extend(direct_sons)
-                    out.write(f'    {" ".join([self._type_name_or_object_freshname(t) for t in direct_sons])} - {self._type_name_or_object_freshname(current_type)}\n')
+                    out.write(f'    {" ".join([self._get_mangled_name(t) for t in direct_sons])} - {self._get_mangled_name(current_type)}\n')
             out.write(' )\n')
         else:
-            out.write(f' (:types {" ".join([cast(UT, t).name for t in self.problem.user_types])})\n' if len(self.problem.user_types) > 0 else '')
+            out.write(f' (:types {" ".join([self._get_mangled_name(t) for t in self.problem.user_types])})\n' if len(self.problem.user_types) > 0 else '')
+
+        # Iterate the actions to retrieve domain objects
+        for a in self.problem.actions:
+            if isinstance(a, up.model.InstantaneousAction):
+                for p in a.preconditions:
+                    self.domain_objects = self.domain_objects | obe.get(p)
```

Comment:
this way of adding to a set is very costly because you create a new set, you can simply add to the set in-place with `|=`
```suggestion
                    self.domain_objects |= obe.get(p)
```